crates/containerd-shim-wasm/src/sandbox/cli.rs

```diff
@@ -199,69 +227,24 @@ pub fn shim_main<'a, I>(
         std::process::exit(0);
     }
 
+    let config = config.unwrap_or_default();
+
     // Initialize the zygote and logger for the container process
     #[cfg(unix)]
-    {
-        let default_config = Config::default();
-        let config = config.as_ref().unwrap_or(&default_config);
-        init_zygote_and_logger(flags.debug, config);
-    }
+    init_zygote_and_logger(flags.debug, &config);
 
-    #[cfg(feature = "opentelemetry")]
-    if otel_traces_enabled() {
-        // opentelemetry uses tokio, so we need to initialize a runtime
-        async {
-            let otlp_config = OtlpConfig::build_from_env().expect("Failed to build OtelConfig.");
-            let _guard = otlp_config
-                .init()
-                .expect("Failed to initialize OpenTelemetry.");
-            tokio::task::block_in_place(move || {
-                shim_main_inner::<I>(name, version, revision, shim_version, config);
-            });
-        }
-        .block_on();
-    } else {
-        shim_main_inner::<I>(name, version, revision, shim_version, config);
-    }
+    async {
+        let _guard = otel_traces_guard();
+        init_traces_context();
 
-    #[cfg(not(feature = "opentelemetry"))]
-    {
-        shim_main_inner::<I>(name, version, revision, shim_version, config);
+        let shim_version = shim_version.into().unwrap_or("v1");
+        let lower_name = name.to_lowercase();
+        let shim_id = format!("io.containerd.{lower_name}.{shim_version}");
+
+        run::<ShimCli<I>>(&shim_id, Some(config)).await;
     }
+    .block_on();
 
     #[cfg(target_os = "linux")]
     log_mem();
 }
-
-#[cfg_attr(feature = "tracing", tracing::instrument(level = "Info"))]
-fn shim_main_inner<'a, I>(
```

Comment:
I tried a similar refactoring before, but I realized that if there is only one `shim_main` function, then it's own trace will not be exported to the trace collector becuase this function initializes the opentelemetry infras. This is the main reason I splitted it to `shim_main` and `shim_main_inner` so that one can initialize a otel runtime and another one can export it's function tracing to the collector. 

I am assuming with your refactoring we will not be able to see `shim_main` trace anymore.

Reply:
oh, good catch, thanks!